app/org/sagebionetworks/bridge/cache/CacheProvider.java

```diff
@@ -26,41 +34,72 @@ public void setJedisOps(JedisOps jedisOps) {
+            throw new BridgeServiceException(e);
+        }
+        if (results == null) {
+            throw new BridgeServiceException("Session storage error.");
+        }
+        for (Object result : results) {
             if (!"OK".equals(result)) {
-                throw new BridgeServiceException("Session storage error");
+                throw new BridgeServiceException("Session storage error.");
             }
-        } catch (Throwable e) {
-            promptToStartRedisIfLocalEnv(e);
-            throw new BridgeServiceException(e);
         }
     }
 
-    public UserSession getUserSession(String key) {
+    public UserSession getUserSession(final String sessionToken) {
         try {
-            String redisKey = RedisKey.SESSION.getRedisKey(key);
-            String ser = jedisOps.get(redisKey);
+            final String sessionKey = RedisKey.SESSION.getRedisKey(sessionToken);
+            final String ser = jedisOps.get(sessionKey);
             if (ser != null) {
-                jedisOps.expire(redisKey, BridgeConstants.BRIDGE_SESSION_EXPIRE_IN_SECONDS);
-                return BridgeObjectMapper.get().readValue(ser, UserSession.class);
+                final UserSession session = BridgeObjectMapper.get().readValue(ser, UserSession.class);
+                final String userKey = RedisKey.USER_SESSION.getRedisKey(session.getUser().getId());
+                jedisOps.getTransaction(sessionToken)
+                        .expire(userKey, BridgeConstants.BRIDGE_SESSION_EXPIRE_IN_SECONDS)
+                        .expire(sessionKey, BridgeConstants.BRIDGE_SESSION_EXPIRE_IN_SECONDS)
+                        .exec();
+                return session;
             }
+            return null;
         } catch (Throwable e) {
-            promptToStartRedisIfLocalEnv(e);
+            promptToStartRedisIfLocal(e);
             throw new BridgeServiceException(e);
         }
-        return null;
     }
 
-    public void removeSession(String key) {
+    public void removeSession(final UserSession session) {
         try {
-            String redisKey = RedisKey.SESSION.getRedisKey(key);
-            jedisOps.del(redisKey);
+            final String sessionKey = RedisKey.SESSION.getRedisKey(session.getSessionToken());
+            final String userKey = RedisKey.USER_SESSION.getRedisKey(session.getUser().getId());
+            jedisOps.getTransaction().del(sessionKey).del(userKey).exec();
+        } catch(Throwable e) {
+            promptToStartRedisIfLocal(e);
+            throw new BridgeServiceException(e);
+        }
+    }
+
+    public void removeSessionByUserId(final String userId) {
+        try {
+            final String userKey = RedisKey.USER_SESSION.getRedisKey(userId);
+            final String sessionToken = jedisOps.get(userKey);
+            final String sessionKey = RedisKey.SESSION.getRedisKey(sessionToken);
+            jedisOps.getTransaction().del(sessionKey).del(userKey).exec();
```

Comment:
What if the user is not logged in? I believe this will actually crash with a NullPointerException.